lib/Dialect/SV/Ops.cpp

```diff
@@ -90,9 +90,143 @@ static void printAlwaysAtPosEdgeOp(OpAsmPrinter &p, AlwaysAtPosEdgeOp op) {
 }
 
 //===----------------------------------------------------------------------===//
+// Structure operations
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// InterfaceOp
+
+static ParseResult parseInterfaceOp(OpAsmParser &parser,
+                                    OperationState &result) {
+  StringAttr nameAttr;
+  if (parser.parseSymbolName(nameAttr, ::mlir::SymbolTable::getSymbolAttrName(),
+                             result.attributes))
+    return failure();
+
+  Region *body = result.addRegion();
+  if (parser.parseRegion(*body, llvm::None, llvm::None))
+    return failure();
+
+  InterfaceOp::ensureTerminator(*body, parser.getBuilder(), result.location);
+
+  return success();
+}
+
+static void printInterfaceOp(OpAsmPrinter &p, InterfaceOp op) {
+  p << op.getOperationName() << ' ';
+  p.printSymbolName(op.getName());
+  p.printRegion(op.body(),
+                /*printEntryBlockArgs=*/false,
+                /*printBlockTerminators=*/false);
+}
+
+//===----------------------------------------------------------------------===//
+// InterfaceSignalOp
+
+static ParseResult parseInterfaceSignalOp(OpAsmParser &parser,
+                                          OperationState &result) {
+  StringAttr nameAttr;
+  TypeAttr type;
+  if (parser.parseSymbolName(nameAttr, ::mlir::SymbolTable::getSymbolAttrName(),
+                             result.attributes) ||
+      parser.parseColon() ||
+      parser.parseAttribute(type, "type", result.attributes))
+    return failure();
+
+  return success();
+}
+
+static void printInterfaceSignalOp(OpAsmPrinter &p, InterfaceSignalOp op) {
```

Comment:
Use custom format rather than C++ printer/parser for this one.

Reply:
Custom assembly format sounds good to me. I'll update this PR. I was thinking the custom printer/parser might lead to a slightly more terse textual representation, but it is probably more trouble than its worth...

Reply:
As long as you don't have a region, I don't see much reason to use a c++ parser/printer.  You should be able to get the same syntax, I think.

Reply:
@stephenneuendorffer updated. I think I see two differences between the customer parser/printer and the declarative assembly format:

1. For both the `InterfaceSignalOp` and `InterfaceModportOp`, the symbol name attribute doesn't go through `parseSymbolName`/`printSymbolName`, so it looks like a regular `"string"` attribute instead of the nice `@symbol` way of printing/parsing
2. I didn't see a way to format the array of structs in the InterfaceModportOp other than explicitly writing out the dictionaries with keywords and values. This is more verbose, but I think it is nevertheless a clear format to read/test against/etc.

I may be missing something. Not sure if there are any further knobs to twist to tweak the parsing and printing of these attributes.

To me, it seems more valuable to use the declarative style rather than writing imperative C++ to address the above. I'll take any suggestions on how to further improve the assembly format, but I'm happy with how it is now.

Reply:
@River707 Do you have any suggestions here?

Reply:
Hmmm, I sent out a few things for review that could probably help here:

* D84719 adds support for "custom" directives, which allows for injecting custom parsing for specific parts of an otherwise declarative format. (e.g. the handling for "ports")
* D86759 adds support for formatting symbol names properly in the declarative format
* D86760 adds some support for regions (those without custom arg handling), and if you need more complex support you could use a custom directive.

Reply:
I'm looking forward to D84719!

Reply:
Thanks @River707 ! Looking forward to using these new features here in CIRCT.